contracts/CarbonVoteX.sol

```diff
@@ -0,0 +1,178 @@
+    event _SendGas(address indexed _msgSender, bytes32 _pollId);
+    
+    struct Poll {
+        // voter address => choice => votes
+        // choice can be hashed value of voting options
+        // e.g. sha3("choice #1"), sha3("choice #2")
+        mapping (address => mapping(bytes32 => uint)) voterVotes;
+        // choice => votes
+        mapping (bytes32 => uint) choiceVotes;
+        // voter address => available votes;
+        mapping (address => uint) availableVotes;
+        //map voter's address to the amount of gas sent
+        mapping (address => uint) gasSentByVoter;
+        uint startBlock;
+        uint endBlock;
+        bytes32 pollId;
+        // the ERC20 token used
+        ERC20 token;
+    }
+
+    // associative container of staking events
+    // _pollId => PollInfo
+    mapping (bytes32 => Poll) public polls;
+
+    // master's address
+    address private master;
+
+    /*
+     * Modifiers
+     */
+    modifier isMaster(){
+        require (msg.sender == master);
+        _;
+    }
+
+    // @param _pollId the id of poll
+    // returns the startBlock, endBlock, pollId token address of the poll.
+    function getPoll (bytes32 _pollId) public view returns (uint _startBlock, uint _endBlock, bytes32 _rpollId, address _tokenAddr){
+        //Check if poll exists
+        require(pollExist(_pollId));
+
+        Poll storage poll = polls[_pollId];
+        return (poll.startBlock, poll.endBlock, poll.pollId, poll.token);
+    }
+
+    // @param _master address which is able to write votes
+    // the master address is owned by a backend server
+    constructor (address _master) public{
+        master = _master;
+    }
+
+    // @param _voter the address of the voter
+    // returns the amount of gas remining for _voter
+    function getGasSent(bytes32 _pollId, address _voter) public view returns (uint){
+        //Check if poll exists
+        require(pollExist(_pollId));
+
+        return polls[_pollId].gasSentByVoter[_voter];
+    }
+
+    // store the amount of gas sent by the voter 
+    function sendGas(bytes32 _pollId) public payable {
+        // poll must exits and not yet expired. 
+        require (pollExist(_pollId) && !pollExpired(_pollId));
+
+        emit _SendGas(msg.sender, _pollId);
+        polls[_pollId].gasSentByVoter[msg.sender] += msg.value;
+    }
+
+    // @param _pollId UUID (hash value) of a poll
+    // @param _startBlock starting block (unix timestamp) of the event
+    // @param _endBlock ending block (unix timestamp) of the event
+    // register a new poll
+    // Note that we do not allow re-registrations for the same UUID
+    function register(bytes32 _pollId, uint _startBlock, uint _endBlock, address _tokenAddr) public {
+        //Check if poll exists.
+        require (pollExist(_pollId) == false);
+        // check resonable endBlock and startBlock
+        require (_endBlock > block.number && _startBlock < _endBlock);
```

Comment:
shouldn't we check block.number < startBlock instead of _endBlock > block.number?